### PR TITLE
Closes #29 - Save all programs that users run

### LINKED ISSUE #29: Save all "play"ed and "submit"ted LP programs associated with a token
To better support users, it would be nice if every version of a bot program that users 'play' or 'submit' would be saved with a filename including the hash of the file. The programs could potentially be compressed, but LP programs are pretty small in general. We could also use this to recover old version if necessary, but it would be really nice for debugging, along with the next enhancement.

### PR DESCRIPTION
**Pull request by derpferd**
This PR changes how and when users' programs are saved.

Original logic:
 - Every time a user clicks `submit` the current user program is sent to the server.
 - The server compiles and runs the user program.
 - The server saves the program for the user overwriting previous program.
 - When scoring programs the db simply reads the code file.

New logic
 - Each time a user hits the `play` or `submit` button program is sent to the server.
 - The server compiles and runs the user program.
 - The server saves the program using a timestamp and hash in a new directory.
 - If the request was a submit request the server marks the newly saved program as the `active` program.
 - When scoring programs the db finds which code is active then reads from that directory.

The feature by itself isn't user facing but it enables many new cool user facing features like #30, viewing older versions, allowing users to save multiple programs and selecting the one with the highest score, etc...

### CODE REVIEW

File: CYLGame/www/templates/grid.html [hunk 1]
```diff
@@ -136,11 +136,17 @@
...
+                token_value = "";
+            } else {
+                token_value = localStorage.user_token;
+            }
             $("#debugTable").html("");
             $.ajax({
                 type: "POST",
                 url: $SCRIPT_ROOT + 'sim',
-                data: JSON.stringify({code: editor.getValue(), seed: window.seed, opponents: opponents}),
+                data: JSON.stringify({code: editor.getValue(), token:token_value, seed: window.seed, opponents: opponents}),
```
Comment (ChandlerSwift) [inline CYLGame/www/templates/grid.html:149]: Nit: Do you want to add a space after the colon here?

**Review by ChandlerSwift — DISMISSED** [review]

File: CYLGame/www/templates/grid.html (hunk 1 above)
Comment (derpferd) [reply CYLGame/www/templates/grid.html:149]: Yes I do :)

**Review by ChandlerSwift — DISMISSED** [review]
LGTM. Few nitpicks, if you're so inclined. 

Aside: took me awhile to figure out why this looked so familar. 

> ChandlerSwift reviewed on Nov 3, 2020

ah, yes :joy:

**Comment by derpferd** [comment]
Retry DeepCode

**Review by ChandlerSwift — APPROVED** [review]
I've run out of nits to pick! This looks great. Thanks!

**Pull request merged**